aiohttp/connector.py

```diff
@@ -69,6 +72,76 @@
     from .tracing import Trace
 
 
+def _convert_hosts_to_addr_infos(
+    hosts: List[Dict[str, Any]]
+) -> List[aiohappyeyeballs.AddrInfoType]:
+    """Converts the list of hosts to a list of addr_infos.
+
+    The list of hosts is the result of a DNS lookup. The list of
+    addr_infos is the result of a call to `socket.getaddrinfo()`.
+    """
+    addr_infos: List[aiohappyeyeballs.AddrInfoType] = []
+    for hinfo in hosts:
+        host = hinfo["host"]
+        is_ipv6 = helpers.is_ipv6_address(host)
+        family = socket.AF_INET6 if is_ipv6 else socket.AF_INET
+        addr = (host, hinfo["port"], 0, 0) if is_ipv6 else (host, hinfo["port"])
+        addr_infos.append(
+            (family, socket.SOCK_STREAM, socket.IPPROTO_TCP, hinfo["hostname"], addr)
+        )
+    return addr_infos
+
+
+def _convert_local_addr_to_addr_infos(
+    local_addr: Optional[Tuple[str, int]]
+) -> Optional[List[aiohappyeyeballs.AddrInfoType]]:
+    """Convert a local_addr tuple to a list of addr_infos."""
+    if local_addr is None:
+        return None
+    host, port = local_addr
+    is_ipv6 = helpers.is_ipv6_address(host)
+    family = socket.AF_INET6 if is_ipv6 else socket.AF_INET
+    addr = (host, port, 0, 0) if is_ipv6 else (host, port)
+    return [(family, socket.SOCK_STREAM, socket.IPPROTO_TCP, "", addr)]
+
+
+def _pop_addr_infos_interleave(
+    addr_infos: List[aiohappyeyeballs.AddrInfoType], interleave: int
+) -> None:
+    """Pop addr_info from the list of addr_infos by family up to interleave times.
+
+    The interleave parameter is used to know how many addr_infos for
+    each family should be popped of the top of the list.
+    """
+    seen: Dict[int, int] = {}
+    to_remove: List[aiohappyeyeballs.AddrInfoType] = []
+    for addr_info in addr_infos:
+        family = addr_info[0]
+        if family not in seen:
+            seen[family] = 0
+        if seen[family] < interleave:
+            to_remove.append(addr_info)
+        seen[family] += 1
+    for addr_info in to_remove:
+        addr_infos.remove(addr_info)
+
+
+def _remove_addr_infos(
+    addr_infos: List[aiohappyeyeballs.AddrInfoType],
+    addr: Tuple[str, ...],
+) -> None:
+    """Pop addr_info from the list of addr_infos by addr."""
+    bad_addrs_infos: List[aiohappyeyeballs.AddrInfoType] = []
+    for addr_info in addr_infos:
+        if addr_info[-1][0] == addr[0]:
+            bad_addrs_infos.append(addr_info)
+    if bad_addrs_infos:
+        for bad_addr_info in bad_addrs_infos:
+            addr_infos.remove(bad_addr_info)
+    else:
+        addr_infos.pop(0)
```

Comment:
This should probably raise instead of popping the first one since it means we can't find the bad peer for some reason.  Likely a problem comparing the addresses since the str may be formatted different. 

Need to add coverage for this